File: backend/src/Operum.API/Controllers/TrackerConstantsController.cs
using Microsoft.AspNetCore.Mvc;
using Operum.API.Controllers.Base;
using Operum.Model.DTOs.TrackerConstants.Requests;
using Operum.Service.Interfaces;

namespace Operum.API.Controllers
{
    [ApiController]
    [Route("api/trackers/{trackerId}/constants")]
    public class TrackerConstantsController(ITrackerConstantsService trackerConstantsService) : BaseController
    {
        [HttpGet]
        public async Task<IActionResult> GetConstantList([FromRoute] string trackerId)
        {
            return GetApiResponse(await trackerConstantsService.GetConstantList(trackerId));
        }

        [HttpGet("{constantId}")]
        public async Task<IActionResult> GetConstant([FromRoute] string trackerId, [FromRoute] string constantId)
        {
            return GetApiResponse(await trackerConstantsService.GetConstant(trackerId, constantId));
        }

        [HttpPost]
        public async Task<IActionResult> CreateConstant([FromRoute] string trackerId, [FromBody] CreateTrackerConstantDto dto)
        {
            return GetApiResponse(await trackerConstantsService.CreateConstant(trackerId, dto));
        }

        [HttpPut("{constantId}")]
        public async Task<IActionResult> UpdateConstant([FromRoute] string trackerId, [FromRoute] string constantId, [FromBody] UpdateTrackerConstantDto dto)
        {
            return GetApiResponse(await trackerConstantsService.UpdateConstant(trackerId, constantId, dto));
        }

        [HttpDelete("{constantId}")]
        public async Task<IActionResult> DeleteConstant([FromRoute] string trackerId, [FromRoute] string constantId)
        {
            return GetApiResponse(await trackerConstantsService.DeleteConstant(trackerId, constantId));
        }
    }
}

File: frontend/src/features/constants/api/trackerConstantsController.ts
import api from "../../../shared/api/api";
import { ApiResponse } from "../../../shared/types/ApiResponse";
import {
    CreateTrackerConstantDto,
    TrackerConstantDto,
    UpdateTrackerConstantDto,
} from "../types/TrackerConstantDto";

export const trackerConstantsController = {
    getConstants: async (
        trackerId: string
    ): Promise<ApiResponse<TrackerConstantDto[]>> => {
        return await api.get(`/trackers/${trackerId}/constants`);
    },
    createConstant: async (
        trackerId: string,
        values: CreateTrackerConstantDto
    ): Promise<ApiResponse<TrackerConstantDto>> => {
        return await api.post(`/trackers/${trackerId}/constants`, values);
    },
    updateConstant: async (
        trackerId: string,
        constantId: string,
        values: UpdateTrackerConstantDto
    ): Promise<ApiResponse<TrackerConstantDto>> => {
        return await api.put(
            `/trackers/${trackerId}/constants/${constantId}`,
            values
        );
    },
    deleteConstant: async (
        trackerId: string,
        constantId: string
    ): Promise<ApiResponse> => {
        return await api.delete(
            `/trackers/${trackerId}/constants/${constantId}`
        );
    },
};
